[WIP] Refactor file searching

MattWis opened the pull request:
First of all, it's worth noting that none of the tests work, unless you comment out the `os_breakcheck();` in `vim_findfile`. Is there a reason the event loop would behave oddly during testing?

```
vim_findfile_init
```

This function only gets called in two places (`find_file_in_path_option` and `tags.c`). Those calls shared many of the same arguments, so I got rid of arguments that were common between them, and some logic associated with them. (Created the stopdir string in the function instead of passing it in, set maximum depth to 100 as a compile-time constant, always reuse visited lists.)

I also extracted `vim_create_stopdirs` to create a list of stopdirs from the stopdirs string.

Finally, I pulled the structs defined at the beginning of the file into the header file instead. This made testing easier (possible?), and allowed me to actually type the arguments to `vim_findfile_init` and `vim_findfile`.

```
find_file_in_path_option
```

I removed the `goto theend;` statements in favor of an early return, since there wasn't any cleanup logic at `theend`.

I extracted `search_with_suffixes`, which tries to add parts of the list of suffixes to a file and find it that way. 

But mainly I just unrolled awkwardly laid out control structures. Most notably, removing `did_findfile_init`, turning:

```
int did_findfile_init = FALSE;
for(;;) {
  if (did_findfile_init) {
    //search for file
    did_findfile_init = FALSE
  } else {
    if (condition) {
      //cleanup
      break;
    }
    //call vim_findfile_init
    did_findfile_init = TRUE;
  }
}
```

to

```
while(!condition) {
  //call vim_findfile_init
  //search for file
}
//cleanup
```

Testing

To test, I looked for the spec file that I had written. Is that a decent method? I could alternatively create files in a temporary directory. 

It's also not cross-platform yet. (Is there a `get_path_seperator()` in lua?) 

I use the lua filesystem (lfs) package.

justinmk commented:
See also https://github.com/neovim/neovim/pull/1043 and https://github.com/neovim/neovim/pull/1094

I haven't looked at your PR. If you can explain the value, provide thorough tests, and figure out the cross-platform details, it would be a welcome change. But otherwise, the trend tends to be that people send half-implemented refactorings and lose interest because others don't have the time to spell out the details. Unfortunately you may have to be self-motivated for this type of work (though it is appreciated)!

MattWis commented:
Just to clarify: Is "this change makes this area of the codebase more readable and adds tests" an acceptable way to explain the value? (Put another way, if I put in the work to get this change the rest of the way, will it get merged in? Or dismissed for not adding enough value to merge in?)

elmart commented:
Please add description of performed refactorings to commit message, so that others can more easily review this.

splinterofchaos commented:
I looked at the code a bit--there's a lot going on and large portions of code moved, making it hard to follow...  the `:find` utility, in particular, is one huge state machine that has odd interactions with vim's overall state. Look at the regression test, `src/nvim/testdir/test73.in`. It might be a more important validator of these changes than the unit since it is state aware.

> always reuse visited lists

Doesn't that mean that if I `:find file1` and then `:find file2`, having already searched the directory, vim won't find file2?

Travis found a [memory leak](https://travis-ci.org/neovim/neovim/jobs/51728990). The [error on the unit test](https://travis-ci.org/neovim/neovim/jobs/51728989) is odd..., 

```
Error → ./test/unit/helpers.lua @ 113
/home/travis/build/neovim/neovim/test/unit/file_search_spec.lua
./test/unit/helpers.lua:113: declaration specifier expected near 'FileID' at line 2
stack traceback:
	./test/unit/helpers.lua:113: in function 'cimport'
	...ravis/build/neovim/neovim/test/unit/file_search_spec.lua:10: in main chunk
```

It's hard to tell if #1043 was really a great idea since it naively modified the internal state, and when I attempted to write the regression test for upstream, either I got the test wrong or it had a bug. Bram's comment was that `:find` has been the source of many bugs in the past, so changes to it must be made very carefully. Please, if possible, break up the changes into more granular steps. The changes you've listed seem mostly logical enough, but it's hard to know 100%.

justinmk commented:
> Or dismissed for not adding enough value to merge in? 

As far as I'm concerned any change that is covered thoroughly by tests is almost guaranteed to be merged, if the change is at least somewhat beneficial. Your changes do sound valuable. 

The tensions about refactoring some of these key pieces of the code are:

- Vim has some very specific (and sometimes stupid) quirks that must be preserved. Refactoring without generous test coverage just causes bugs in an area of the code that no one "owns". 
    - @splinterofchaos is perhaps most familiar with this area, so I'm glad he chimed in!
- If bugs are fixed upstream, we won't get those fixes for "free" anymore, so the change had better be worth it :)
- Need to take #1043 into account if @splinterofchaos thinks it is still a possibility.

> did_findfile_init

Why was Vim caching that, and how do your changes make it unnecessary?

justinmk commented:
> Is there a reason the event loop would behave oddly during testing?

Yes, there is no event loop for the unit tests. You can spin the event loop manually, I think, or you can set globals manually. If it gets too crazy, you could write "functional" tests instead of unit tests (though this PR should have *some* unit tests for granularity).

> To test, I looked for the spec file that I had written. Is that a decent method? I could alternatively create files in a temporary directory.

Up to you.

> It's also not cross-platform yet.

Do you mean the tests only?

splinterofchaos commented on src/nvim/file_search.c:
```diff
@@ -1453,44 +1330,32 @@ find_file_in_path_option (
+      if (fdip_search_ctx != NULL) {
+        file_name = vim_findfile(fdip_search_ctx);
+        if (file_name != NULL)
+          return file_name;
```
Returning here leaves `buf` unfreed.

splinterofchaos commented on src/nvim/file_search.c:
```diff
@@ -1509,7 +1374,25 @@ find_file_in_path_option (
+  return NULL;
 }
 
+char_u *search_with_suffixes(char_u *suffixes, size_t l, int find_what) {
```
This function should be marked `static`. Can `suffixes` be `const char_u *`? Please change `l` to `len` for readability.

splinterofchaos commented:
> Need to take #1043 into account if @splinterofchaos thinks it is still a possibility.

Shouldn't be a problem, actually. The biggest conflict is that I renamed `ff_stack` to `ff_state`. but that won't be hard to resolve. I might be able to dedicate some time this week for reviving an old PR or two, but that won't be the one.

MattWis commented:
In order of comments:

@elmart: I'm going to break the commit up so that they make more sense.

@splinterofchaos: It will find both files. (At least it did for me.) Each time you call find, a new search context is initialized. (I believe.)

And failing on importing the header is a really weird error. I'll look into it.

@justinmk: Good to know that it provides value (if I deliver).

The control structure for using ```did_findfile_init``` was the same as a very-convoluted while loop. The same commands should now be run in the same order, but the while loop makes it much clearer what the intended result is. There wasn't really any caching, since it was a local variable initialized to ```FALSE```, and never changed until the ```for(;;)``` loop.

And yes, only the tests are not cross-platform. (Should have made that clearer.)

@splinterofchaos: I'll look through to see what else I can mark as static, and I feel pretty stupid that there's a one letter variable name in there. 

If I get this commit in a decent state, I can look into merging in #1043.

MattWis committed:
Add tests for file searching

To do this, the structs from file_search.c had to be moved into
file_search.h, along with typing the ff_search_ctx_T arguments of
vim_findfile_init and vim_findfile.

MattWis committed:
Remove parameters from vim_findfile_init

The only two places vim_findfile_init is called in the codebase are in
the convenience function find_file_in_path_option, and in tags.c. Both
of those places used 3 arguments that were the same, so they were
removed as arguments.

char_u *stopdirs: This argument was always created with the
vim_findfile_stopdir function, so we now do that inside of
vim_findfile_init.

int level: Sets the maximum number of levels to recurse into during
vim_findfile. Was always 100, and is now set in file_search.h

int free_visited: This determines whether or not to reuse old visited
lists. Was always FALSE, so that removes one branch in
vim_findfile_init.

MattWis committed:
Refactor: vim_create_stopdirs

This helper function populates the ffsc_stopdirs_v argument of the
search context in vim_findfile_init.

MattWis committed:
Add assert for find_what argument

In vim_findfile_init, find_what should be one of 3 #define'd constants:
FINDFILE_FILE, FINDFILE_DIR, or FINDFILE_BOTH. Check that is the case
with an assertion.

MattWis committed:
Refactor: populate_wc_path

This helper function populates the ffsc_wc_path member of the search
context in vim_findfile_init, if there is a wildcard in the path.

MattWis committed:
Refactor part of find_file_in_path_option

This control structure is really a while loop. It was masquerading as an
infinite for loop that broke upon meeting a condition. Also, by
swtiching the order of the code a bit, it becomes clear that
did_findfile_init isn't necessary at all.

MattWis committed:
Refactor search_with_suffixes, file_of_type_exists

The function search_with_suffixes  searches by trying to add a list of
suffixes to a file name one at a time. Also, this commit unrolls a for
loop that happened either once or twice based on a conditional.

The function file_of_type_exists is a conditional check that the file
exists, and is of the type that we are looking for.

Also, favor early return over goto, removing the need for the
"file_name == NULL" check for the errors

mseeber commented on src/nvim/file_search.c:
```diff
@@ -208,9 +91,10 @@ static char_u e_pathtoolong[] = N_("E854: path too long for completion");
  *
  * Find the file 'filename' in the directory 'path'.
  * The parameter 'path' may contain wildcards. If so only search 'level'
```
outdated comment, 'level' was removed from siganture

mseeber commented:
There are some more outdated comments/descriptions on functions you have changed.

MattWis committed:
Update file_search.c comments

Get rid of all references to 'levels', since we are no longer using
that. Add comments describing the role of the extracted helper
functions.

MattWis committed:
Test vim_create_stopdirs

MattWis committed:
Mark functions static where applicable

MattWis committed:
Make vim_create_stopdirs static

MattWis committed:
Make vim_findfile_stopdir static

justinmk commented:
Not enough test coverage, and there doesn't seem to be enough interest in this part of the code for a drastic refactoring.
